flink-core/src/main/java/org/apache/flink/core/classloading/MutableURLClassLoader.java

```diff
@@ -0,0 +1,152 @@
+ * {@link Configuration}, default order is child-first.
+ */
+public class MutableURLClassLoader extends URLClassLoader {
+
+    private static final Logger LOG = LoggerFactory.getLogger(MutableURLClassLoader.class);
+
+    private final Set<URL> registeredUrls;
+    private final Configuration config;
+    private final ClassLoader parentClassLoader;
+    private volatile FlinkUserCodeClassLoader ownerClassLoader;
+
+    public MutableURLClassLoader(URL[] urls, ClassLoader parent, Configuration config) {
+        super(new URL[0], parent);
+        this.registeredUrls = new HashSet<>(Arrays.asList(urls));
+        this.config = config;
+        this.parentClassLoader = parent;
+        this.ownerClassLoader = buildUserCodeClassLoader();
+    }
+
+    private FlinkUserCodeClassLoader buildUserCodeClassLoader() {
+        final String[] alwaysParentFirstLoaderPatterns =
+                CoreOptions.getParentFirstLoaderPatterns(config);
+
+        // According to the specified class load order，child-first or parent-first
+        // child-first: load from this classloader firstly, if not found, then from parent
+        // parent-first: load from parent firstly, if not found, then from this classloader
+        final String classLoaderResolveOrder =
+                config.getString(CoreOptions.CLASSLOADER_RESOLVE_ORDER);
+        FlinkUserCodeClassLoaders.ResolveOrder resolveOrder =
+                FlinkUserCodeClassLoaders.ResolveOrder.fromString(classLoaderResolveOrder);
+        return (FlinkUserCodeClassLoader)
+                FlinkUserCodeClassLoaders.create(
+                        resolveOrder,
+                        registeredUrls.toArray(new URL[0]),
+                        parentClassLoader,
+                        alwaysParentFirstLoaderPatterns,
+                        NOOP_EXCEPTION_HANDLER,
+                        false);
+    }
+
+    @Override
+    public Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
+        try {
+            return ownerClassLoader.loadClass(name, resolve);
+        } catch (ClassNotFoundException cnf) {
+            throw cnf;
+        }
+    }
+
+    @Override
+    public void addURL(URL url) {
+        // add url to registeredUrls
+        registeredUrls.add(url);
+        // add it to ownerClassLoader
+        ownerClassLoader.addURL(url);
+    }
+
+    /**
+     * Here the old ownerClassLoader can't be close directly because of java class loader's
+     * all-inclusive responsibility delegation mechanism. The mechanism is that when a ClassLoder
+     * loads a class, unless another ClassLoder is specified explicitly, the classes that the class
+     * depends on and references are also loaded by the ClassLoder. Hence, we can't close the old
+     * ownerClassLoader when remove url, it may cause a {@link ClassNotFoundException} exception.
+     *
+     * <p>If call this method frequently, it maybe results in a classloader leak. Please use this
```

Comment:
I'm a bit concerned about the side effect. And I do not fully understand the benefit of the removal. 
What problem would happen if URLs are not removed from a userClassLoader? If it is only for table environment to support multiple different jobs submission from SQL client/gateway, I would suggest to move the class to table module.

Reply:
Yes, you are right. I agree with you, we should move it to table module.